internal/extension/registry/extension_manager.go

```diff
@@ -0,0 +1,183 @@
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/envoyproxy/gateway/api/config/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/envoygateway"
+	"github.com/envoyproxy/gateway/internal/envoygateway/config"
+	"github.com/envoyproxy/gateway/internal/extension/types"
+	"github.com/envoyproxy/gateway/proto/extension"
+)
+
+type hookConn struct {
+	extensionName v1alpha1.ExtensionId
+	hookType      types.HookType
+}
+
+var _ types.Manager = (*Manager)(nil)
+
+type Manager struct {
+	k8sClient     k8scli.Client
+	namespace     string
+	extensions    map[v1alpha1.ExtensionId]v1alpha1.Extension
+	hookConnCache map[hookConn]*grpc.ClientConn
+}
+
+// NewManager returns a new Manager
+func NewManager(cfg *config.Server) (types.Manager, error) {
+	cli, err := k8scli.New(k8sclicfg.GetConfigOrDie(), k8scli.Options{Scheme: envoygateway.GetScheme()})
+	if err != nil {
+		return nil, err
+	}
+
+	// To keep things simple, we're just going to allow only one extension since it gets messy trying to reconcile
+	// xDS from multiple extensions even though the Manager can handle multiple extensions.
+	//
+	// TODO: have the extensions in EnvoyConfig be a data structure and not a list?
+	if len(cfg.EnvoyGateway.Extensions) > 1 {
+		return nil, fmt.Errorf("More than 1 extension is registered. For the time being, we allow up to 1 extension to be registered")
+	}
+
+	var extTable map[v1alpha1.ExtensionId]v1alpha1.Extension
+	for _, extension := range cfg.EnvoyGateway.Extensions {
+		extTable[extension.Name] = *extension
+	}
+
+	return &Manager{
+		k8sClient:  cli,
+		namespace:  cfg.Namespace,
+		extensions: extTable,
+	}, nil
+}
+
+// HasExtension checks to see whether a given Group and Kind has an
+// associated extension registered for it.
+//
+// TODO: Add Kind to the Extension config in the EnvoyGateway config.
+// For now, assume that Kind is always valid and we'll just validate for the API Group.
+func (m *Manager) HasExtension(g v1beta1.Group, k v1beta1.Kind) (bool, *v1alpha1.ExtensionId) {
+	for _, ext := range m.extensions {
+		for _, apiGroup := range ext.APIGroups {
+			if g == apiGroup {
+				extId := ext.Name
+				return true, &extId
+			}
+		}
+	}
+	return false, nil
+}
+
+// GetXDSHookClient returns an XDS Hook Client for an extension with ID extID
+func (m *Manager) GetXDSHookClient(extID v1alpha1.ExtensionId) (types.XDSHookClient, error) {
+	ctx := context.Background()
+
+	ext, ok := m.extensions[extID]
+	if !ok {
+		return nil, fmt.Errorf("extension %s not found", extID)
+	}
+
+	if ext.Service == nil {
+		return nil, fmt.Errorf("Extension service config is nil")
+	}
+
+	key := hookConn{
+		extensionName: ext.Name,
+		hookType:      types.XDSHook,
+	}
+	conn, cached := m.hookConnCache[key]
+	if !cached {
+		serverAddr := fmt.Sprintf("%s:%d", ext.Service.Host, ext.Service.Port)
+
+		opts, err := setupGRPCOpts(ctx, m.k8sClient, &ext, m.namespace)
+		if err != nil {
+			return nil, err
+		}
+
+		conn, err = grpc.Dial(serverAddr, opts...)
```

Comment:
This is blocking correct?

Reply:
No it's not blocking. By default it will return return immediately and the connection is made in the background usually lazily i.e once it makes the first request then it will create a TCP connection and maintain it in the background and it takes care of keeping it healthy. Not entirely sure about go but that was the case in the C-language implementations

Reply:
Ok, so the RPC call will block trying to get the initial connection then. Sounds good.